SmartDeviceLink/private/SDLPresentAlertOperation.m

```diff
@@ -0,0 +1,466 @@
+        return handler();
+    }
+
+    SDLLogD(@"Uploading images for alert");
+    __weak typeof(self) weakself = self;
+    [self.fileManager uploadArtworks:[artworksToBeUploaded copy] progressHandler:^BOOL(NSString * _Nonnull artworkName, float uploadPercentage, NSError * _Nullable error) {
+        __strong typeof(weakself) strongself = weakself;
+        SDLLogD(@"Uploaded alert images: %@, error: %@, percent complete: %f.2%%", artworkName, error, uploadPercentage * 100);
+        if (strongself.isCancelled) {
+            return NO;
+        }
+
+        return YES;
+    } completionHandler:^(NSArray<NSString *> * _Nonnull artworkNames, NSError * _Nullable error) {
+        if (error != nil) {
+            SDLLogE(@"Error uploading alert images: %@", error);
+        } else {
+            SDLLogD(@"All alert images uploaded");
+        }
+
+        return handler();
+    }];
+}
+
+/// Sends the alert RPC to the module. The operation is finished once a response has been received from the module.
+- (void)sdl_presentAlert {
+    __weak typeof(self) weakSelf = self;
+    [self.connectionManager sendConnectionRequest:self.alertRPC withResponseHandler:^(__kindof SDLRPCRequest * _Nullable request, __kindof SDLRPCResponse * _Nullable response, NSError * _Nullable error) {
+        __strong typeof(weakSelf) strongSelf = weakSelf;
+        if (error != nil) {
+            SDLAlertResponse *alertResponse = (SDLAlertResponse *)response;
+            NSMutableDictionary *alertResponseUserInfo = [NSMutableDictionary dictionary];
+            alertResponseUserInfo[@"error"] = error;
+
+            NSNumber *tryAgainTime = nil;
+            if (alertResponse != nil && alertResponse.tryAgainTime != nil) {
+                tryAgainTime = alertResponse.tryAgainTime;
+            }
+            alertResponseUserInfo[@"tryAgainTime"] = tryAgainTime;
+
+            NSError *alertResponseError = [NSError sdl_alertManager_presentationFailed:alertResponseUserInfo];
+            strongSelf.internalError = alertResponseError;
+        }
+
+        [strongSelf finishOperation];
+    }];
+}
+
+#pragma mark - Cancel
+
+/// Cancels the alert. If the alert has not yet been sent to the module, it will not be sent. If the alert is already presented on the module, the alert will be immediately dismissed. Canceling an already presented alert will only work if connected to modules supporting RPC spec versions 6.0+. On older versions alert will not be dismissed.
+- (void)sdl_cancelInteraction {
+    if (self.isFinished) {
+        SDLLogW(@"This operation has already finished so it can not be canceled.");
+        return;
+    } else if (self.isCancelled) {
+        SDLLogW(@"This operation has already been canceled. It will be finished at some point during the operation.");
+        return;
+    } else if (self.isExecuting) {
+        if ([SDLGlobals.sharedGlobals.rpcVersion isLessThanVersion:[[SDLVersion alloc] initWithMajor:6 minor:0 patch:0]]) {
```

Comment:
Are you talking about if the images/files are being uploaded when `cancel` is called? Do you mean the `isExecuting` check or the `rpcVersion` check?

Reply:
Yeah. It's possible for `isExecuting == true` while the alert is not on the screen if it's still uploading. It shouldn't send a `CancelInteraction` and should just cancel the operation in that case.

Reply:
Is there an easier way to do this than using a flag to track if the alert RPC has been presented? I think we will also have to make sure that the flag is thread safe.

Reply:
I don't think there's an easier way. You could use a state machine (`SDLStateMachine`) instead. If it's just a flag, since it's not under an intense load, we could just make it `atomic` or put the set / get into an `@synchronized`. It wouldn't slow the system down much because it wouldn't be set or accessed often.